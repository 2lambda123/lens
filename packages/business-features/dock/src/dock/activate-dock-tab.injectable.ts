import { getInjectable } from "@ogre-tools/injectable";
import { action } from "mobx";
import activeDockTabIdStateInjectable from "./active-dock-tab-id-state.injectable";

const activateDockTabInjectable = getInjectable({
  id: "activate-dock-tab",

  instantiate: (di) => {
    const activeDockTabId = di.inject(activeDockTabIdStateInjectable);

    return action((tabId: string) => activeDockTabId.set(tabId));
  },
});

export default activateDockTabInjectable;
