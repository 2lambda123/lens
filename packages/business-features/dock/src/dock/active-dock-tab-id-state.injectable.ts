import { getInjectable } from "@ogre-tools/injectable";
import { observable } from "mobx";

const activeDockTabIdStateInjectable = getInjectable({
  id: "active-dock-tab-id",
  instantiate: () => observable.box<string | null>(),
});

export default activeDockTabIdStateInjectable;
