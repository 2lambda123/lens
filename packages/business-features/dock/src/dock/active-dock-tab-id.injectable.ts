import { getInjectable } from "@ogre-tools/injectable";
import { observable } from "mobx";

const activeDockTabIdInjectable = getInjectable({
  id: "active-dock-tab-id",
  instantiate: () => observable.box<string | null>(),
});

export default activeDockTabIdInjectable;
