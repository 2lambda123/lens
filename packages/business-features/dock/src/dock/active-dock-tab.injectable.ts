import { getInjectable } from "@ogre-tools/injectable";
import { computed } from "mobx";
import dockTabsInjectable from "./dock-tabs.injectable";

const activeDockTabInjectable = getInjectable({
  id: "active-dock-tab",

  instantiate: (di) => {
    const dockTabs = di.inject(dockTabsInjectable);

    return computed(() => {
      const [
        activeDockTab = {
          id: "no-active-dock-tab",
          TitleComponent: () => null,
          ContentComponent: () => null,
        },
      ] = dockTabs.get();

      return activeDockTab;
    });
  },
});

export default activeDockTabInjectable;
