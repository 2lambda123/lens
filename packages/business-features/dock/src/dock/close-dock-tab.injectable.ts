import { getInjectable } from "@ogre-tools/injectable";
import { DockTabInState, dockTabStateInjectable } from "./dock-tab-state.injectable";
import activateDockTabInjectable from "./activate-dock-tab.injectable";
import { action } from "mobx";

const closeDockTabInjectable = getInjectable({
  id: "close-dock-tab",

  instantiate: (di) => {
    const dockTabState = di.inject(dockTabStateInjectable);
    const activateDockTab = di.inject(activateDockTabInjectable);

    return action((tabToBeClosed: DockTabInState) => {
      const currentTabs = [...dockTabState.values()];
      const currentIndex = currentTabs.indexOf(tabToBeClosed);

      const previousIndex = currentIndex - 1;

      const tabToBeActivated = currentTabs.at(previousIndex);

      dockTabState.delete(tabToBeClosed);

      console.log(previousIndex, currentTabs); // TODO: Fix

      if (tabToBeActivated) {
        activateDockTab(tabToBeActivated.id);
      }
    });
  },
});

export default closeDockTabInjectable;
