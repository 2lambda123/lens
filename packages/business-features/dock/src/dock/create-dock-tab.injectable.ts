import { getInjectable, getInjectionToken } from "@ogre-tools/injectable";
import { runInAction } from "mobx";
import type { DockTabType } from "../dock-tab-type";
import { dockTabStateInjectable } from "./dock-tab-state.injectable";
import { getRandomIdInjectionToken } from "./get-random-id.injectable";
import activateDockTabInjectable from "./activate-dock-tab.injectable";

export interface CreateDockTabParams {
  type: DockTabType;
}

export type CreateDockTab = ({ type }: CreateDockTabParams) => void;

export const createDockTabInjectionToken = getInjectionToken<CreateDockTab>({
  id: "create-dock-tab-injection-token",
});

const createDockTabInjectable = getInjectable({
  id: "create-dock-tab",

  instantiate: (di) => {
    const dockTabState = di.inject(dockTabStateInjectable);
    const getRandomId = di.inject(getRandomIdInjectionToken);
    const activateDockTab = di.inject(activateDockTabInjectable);

    return ({ type }) => {
      runInAction(() => {
        const newTabId = getRandomId();

        dockTabState.add({
          id: newTabId,
          typeId: type.id,
        });

        activateDockTab(newTabId);
      });
    };
  },

  injectionToken: createDockTabInjectionToken,
});

export default createDockTabInjectable;
