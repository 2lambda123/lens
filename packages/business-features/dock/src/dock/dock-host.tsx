import React from "react";
import { Tab } from "./tab";
import { Div, Span } from "@k8slens/ui-components";

export const DockHost = () => (
  <Div _className={["Dock", { isOpen: true }]}>
    <Div _flexParent={{ centeredVertically: true }} className="tabs-container">
      <Tab>
        <Div _flexParent={{ centeredVertically: true }}>
          <Span _wordWrap>Some title</Span>
        </Div>
      </Tab>

      <Tab>
        <Div _flexParent={{ centeredVertically: true }}>
          <Span _wordWrap>Some other title</Span>
        </Div>
      </Tab>
    </Div>

    <Div
      _flexParent={{ centeredVertically: true }}
      _className={[
        "toolbar gaps box grow",
        {
          "pl-0": true,
        },
      ]}
    >
      <div className="dock-menu box grow"></div>
    </Div>

    <div className={`tab-content`} style={{ flexBasis: 420 }}>
      Some content
    </div>
  </Div>
);
