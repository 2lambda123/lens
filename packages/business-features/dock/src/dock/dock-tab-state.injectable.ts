import { getInjectable } from "@ogre-tools/injectable";
import { observable } from "mobx";

export interface DockTabInState {
  id: string;
  typeId: string;
}

export const dockTabStateInjectable = getInjectable({
  id: "dock-tab-state",
  instantiate: () => observable.set<DockTabInState>(),
});
