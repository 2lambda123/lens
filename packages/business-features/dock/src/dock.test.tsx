import { createContainer, DiContainer } from "@ogre-tools/injectable";
import { renderFor } from "@k8slens/test-utils";
import { DockHost } from "./dock/dock-host";
import React from "react";
import type { RenderResult } from "@testing-library/react";

describe("dock", () => {
  let di: DiContainer;

  beforeEach(() => {
    di = createContainer("some-container");

    // registerFeature(di, dockFeature);
  });

  describe("when rendered", () => {
    let rendered: RenderResult;

    beforeEach(() => {
      const render = renderFor(di);

      rendered = render(<DockHost />);
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });
  });
});
