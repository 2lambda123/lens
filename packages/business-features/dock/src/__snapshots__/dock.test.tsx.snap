// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dock when rendered renders 1`] = `
<body>
  <div>
    <div
      class="Dock isOpen"
    >
      <div
        class="flex align-center tabs-container"
      >
        <div
          class="Tab flex gaps align-center"
        >
          <div
            class="label"
          >
            <div
              class="flex align-center"
            >
              <span
                class="overflow-hidden text-ellipsis"
              >
                Some title
              </span>
            </div>
          </div>
        </div>
        <div
          class="Tab flex gaps align-center"
        >
          <div
            class="label"
          >
            <div
              class="flex align-center"
            >
              <span
                class="overflow-hidden text-ellipsis"
              >
                Some other title
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex align-center toolbar gaps box grow pl-0"
      >
        <div
          class="dock-menu box grow"
        />
      </div>
      <div
        class="tab-content"
        style="flex-basis: 420px;"
      >
        Some content
      </div>
    </div>
  </div>
</body>
`;
